torch/csrc/jit/codegen/cuda/kernel_cache.h

```diff
@@ -13,21 +13,125 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
+// Go through a tensor, and grab it's sizes/strides potentially broadcasted
+struct ExtractSizeStride {
+  std::vector<int64_t> sizes;
+  std::vector<int64_t> strides;
+
+  // TODO: broadcasted_size should be handled in codegen directly instead of at
+  //       the integration limbo.
+  explicit ExtractSizeStride(
+      const at::Tensor& val,
+      c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
+    if (broadcasted_size) {
+      // [Note - broadcast support in integration]
+      // PyTorch follows numpy broadcasting rule.
+      // (https://numpy.org/doc/stable/user/basics.broadcasting.html)
+      //
+      // So in case where the rank of two operators differ, we align them on
+      // the higher dimensions, hence the offset o_dim-b_dim to the index here.
+      int b_dim = (int)broadcasted_size->size();
+      int o_dim = (int)val.dim();
+      TORCH_CHECK(b_dim >= o_dim);
+      for (int i = 0; i < b_dim; i++) {
+        sizes.push_back(broadcasted_size->at(i));
+        int index = i + o_dim - b_dim;
+        if (index < 0) {
+          strides.push_back(0);
+        } else if (val.sizes()[index] == sizes[i]) {
+          strides.push_back(val.strides()[index]);
+        } else {
+          TORCH_CHECK(
+              val.sizes()[index] == 1,
+              "Not compatible dimension size for broadcast");
+          strides.push_back(0);
+        }
+      }
+    } else {
+      auto o_dim = val.dim();
+      for (decltype(val.dim()) i{0}; i < o_dim; i++) {
+        sizes.push_back(val.sizes()[i]);
+        strides.push_back(val.strides()[i]);
+      }
+    }
+  }
+};
+
+class CudaKernel {
+ public:
+  void setFusionPtr(std::unique_ptr<Fusion> fusion) {
+    fusion_ = std::move(fusion);
+  }
+
+  Fusion* getFusionPtr() {
+    return fusion_.get();
+  }
+
+  const Fusion* getFusionPtr() const {
+    return fusion_.get();
+  }
+
+  CUmodule& getModule() {
+    return module_;
+  }
+
+  CUfunction& getFunction() {
+    return function_;
+  }
+
+  int16_t& getDevice() {
```

Comment:
why return a mutable reference - do we want to allow changing it?

Reply:
Yes, there is a place to update it.

I'm using return by reference so we can have consistent way to update/access `device_`/`function_`/`module_`. It looks still messy right now. Let me try again.

Reply:
Just an idea: it may be less surprising to use an explicit setter here: setDevice()

Reply:
Missed this comment. but I happen to use `setDevice()` anyway.
Great minds think alike, so do us. :laughing: